projects/output-and-error-handling/client/main_test.go

```diff
@@ -0,0 +1,167 @@
+
+func TestMakeWeatherRequest_Success(t *testing.T) {
+	expRespSlice := [][]byte{[]byte("Today it will be sunny!"), []byte("Tomorrow it will be rainy!")}
+
+	for i, expected := range expRespSlice {
+		t.Run(fmt.Sprintf("Case %d", i+1), func(t *testing.T) {
+			svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+				w.Write(expected)
+			}))
+			defer svr.Close()
+			client := &http.Client{}
+			resp, err := makeWeatherRequest(svr.URL, client, 3)
+			if err != nil {
+				t.Errorf("expected err to be nil got %v", err)
+			}
+
+			if resp != string(expected) {
+				t.Errorf("expected response body to be %s got %s", expected, resp)
+			}
+		})
+	}
+}
+
+func TestMakeWeatherRequest_InternalServerError(t *testing.T) {
+	errorMessage := "Internal Server Error"
+	errorStatusCode := http.StatusInternalServerError
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(errorStatusCode)
+		w.Write([]byte(errorMessage))
+	}))
+	defer svr.Close()
+
+	client := &http.Client{}
+	_, err := makeWeatherRequest(svr.URL, client, 3)
+	if err == nil {
+		t.Error("expected error, got nil")
+	}
+
+	expectedError := fmt.Sprint(errorStatusCode, " : ", errorMessage)
+	if err.Error() != expectedError {
+		t.Errorf("expected error message to be %s, got %s", expectedError, err.Error())
+	}
+}
+
+func TestHandleRateLimited(t *testing.T) {
+	testCases := []testCase{
+		{
+			retryAfterTime: "3",
+			retries:        2,
+			expectedValue:  3,
+			expectedError:  nil,
+			description:    "Success",
+		},
+		{
+			retryAfterTime: "3",
+			retries:        4,
+			expectedValue:  3,
+			expectedError:  fmt.Errorf("internal Error : Failed to retry due to exceeded rate limit"),
+			description:    "Failure - Exceed rate limit",
+		},
+		{
+			retryAfterTime: "6",
+			retries:        2,
+			expectedValue:  6,
+			expectedError:  fmt.Errorf("internal Error : Failed to retry due to exceeded duration limit"),
+			description:    "Failure - Exceed duration limit",
+		},
+		{
+			retryAfterTime: "1",
+			retries:        2,
+			expectedValue:  1,
+			expectedError:  fmt.Errorf("internal Error : Failed to retry  for an unknown reason"),
+			description:    "Failure - Unknown reason",
+		},
+	}
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("Case %d %s", i+1, tc.description), func(t *testing.T) {
+			err := handleRateLimited(tc.retryAfterTime, tc.retries)
+			if tc.expectedError == nil {
+				if err != nil {
+					t.Errorf("expected err to be nil got %v", err)
+				}
+			} else {
+				if err.Error() != tc.expectedError.Error() {
+					t.Errorf("expected error message to be %s, got %s", tc.expectedError, err.Error())
+				}
+			}
+		})
+	}
+}
+
+func (m MockTimeProvider) Now() time.Time {
+	return time.Date(2024, time.March, 22, 12, 0, 0, 0, time.UTC)
+}
+
+func (m MockTimeProvider) Until(t time.Time) time.Duration {
+	return t.Sub(m.Now())
+}
```

Comment:
I would move these methods so they are next to the struct. There are also libraries to create mocks:
* https://github.com/uber-go/mock
* https://github.com/vektra/mockery

They can be a little hard to get your head around at first, so let me know if you'd like to have a session or some examples

Reply:
I don't think I can use any library yet. I'm still trying to have a better understanding of Go basic concepts.